once/_iterator_wrappers.py

```diff
@@ -0,0 +1,163 @@
+                    return
+                else:
+                    # We already have the correct result, so we grab it here to
+                    # yield it outside the lock.
+                    next_val = self.results[i]
+
+            if generating:
+                try:
+                    next_val = await generating
+                except StopAsyncIteration:
+                    async with self.lock:
+                        self.generator = None  # Allow this to be GCed.
+                        self.finished = True
+                        self.generating = None
+                        generating = None
+                        return
+                async with self.lock:
+                    self.results.append(next_val)
+                    generating = None
+                    self.generating = None
+
+            send = yield next_val
+            i += 1
+
+
+class _IteratorAction(enum.Enum):
+    # Generating the next value from the underlying iterator
+    GENERATING = 1
+    # Yield an already computed value
+    YIELDING = 2
+    # Waiting for the underlying iterator, already triggered from another call.
+    WAITING = 3
+
+
+class GeneratorWrapper:
+    """Wrapper around an sync generator which only runs once.
+
+    Subsequent calls will return results from the first call, which is
+    evaluated lazily.
+    """
+
+    def __init__(self, func, *args, **kwargs) -> None:
+        self.generator: collections.abc.Generator | None = func(*args, **kwargs)
+        self.finished = False
+        self.results: list = []
+        self.generating = False
+        self.lock = threading.Lock()
+
+    def yield_results(self) -> collections.abc.Generator:
+        # Fast path for subsequent repeated call:
+        with self.lock:
+            finished = self.finished
+        if finished:
+            yield from self.results
+            return
+        i = 0
+        yield_value = None
+        next_send = None
+        # Fast path for subsequent calls will not require a lock
+        while True:
+            action: _IteratorAction | None = None
+            # With a lock, we figure out which action to take, and then we take it after release.
+            with self.lock:
+                if i == len(self.results):
+                    if self.finished:
+                        return
+                    if self.generating:
+                        action = _IteratorAction.WAITING
+                    else:
+                        action = _IteratorAction.GENERATING
+                        self.generating = True
+                else:
+                    action = _IteratorAction.YIELDING
+                    yield_value = self.results[i]
+            if action == _IteratorAction.WAITING:
+                continue
+            if action == _IteratorAction.YIELDING:
+                next_send = yield yield_value
+                i += 1
+                continue
+            if action != _IteratorAction.GENERATING:
```

Comment:
How is this possible? There are only three options and we already exhausted 2, right?

Reply:
you're right, converted to assertion